lib/VhostContainer.php

```diff
@@ -110,81 +111,73 @@ public function setupHttpDrivers(...$args) {
-        }
-        throw new \Error(
-                "Cannot retrieve default host; no Vhost instances added to the group"
-            );
-    }
+        $addressPortWildcardHost = "*:0{$serverId}";
+        if (isset($this->vhosts[$addressPortWildcardHost])) {
+            return $this->vhosts[$addressPortWildcardHost];
+        }
+
+        if ($client->serverAddr[0] == "/") { // unix domain socket
+            // there is no such thing like interface wildcards for unix domain sockets
+            return null; // nothing found
+        }
 
-    private function selectHostByAuthority(InternalRequest $ireq) {
-        $explicitHostId = "{$ireq->uriHost}:{$ireq->uriPort}";
-        $wildcardHost = "0.0.0.0:{$ireq->uriPort}";
-        $ipv6WildcardHost = "[::]:{$ireq->uriPort}";
+        $wildcardIP = \strpos($client->serverAddr, ":") === false ? "0.0.0.0" : "[::]";
+        $serverId = ":$wildcardIP:{$client->serverPort}";
 
+        $explicitHostId = "{$ireq->uriHost}:{$ireq->uriPort}{$serverId}";
         if (isset($this->vhosts[$explicitHostId])) {
-            $vhost = $this->vhosts[$explicitHostId];
-        } elseif (isset($this->vhosts[$wildcardHost])) {
-            $vhost = $this->vhosts[$wildcardHost];
-        } elseif (isset($this->vhosts[$ipv6WildcardHost])) {
-            $vhost = $this->vhosts[$ipv6WildcardHost];
-        } elseif ($this->cachedVhostCount !== 1) {
-            return null;
-        } else {
-            $ipComparison = $ireq->uriHost;
-
-            if (!@inet_pton($ipComparison)) {
-                $ipComparison = (string) substr($ipComparison, 1, -1); // IPv6 braces
-                if (!@inet_pton($ipComparison)) {
-                    return null;
-                }
-            }
-            if (!(($vhost = $this->getDefaultHost()) && in_array($ireq->uriPort, $vhost->getPorts($ipComparison)))) {
-                return null;
-            }
+            return $this->vhosts[$explicitHostId];
+        }
+
+        $addressWildcardHost = "*:{$ireq->uriPort}{$serverId}";
+        if (isset($this->vhosts[$addressWildcardHost])) {
+            return $this->vhosts[$addressWildcardHost];
+        }
+
+        $portWildcardHostId = "{$ireq->uriHost}:0{$serverId}";
+        if (isset($this->vhosts[$portWildcardHostId])) {
+            return $this->vhosts[$portWildcardHostId];
         }
 
-        // IMPORTANT: Wildcard IP hosts without names that are running both encrypted and plaintext
-        // apps on the same interface (via separate ports) must be checked for encryption to avoid
-        // displaying unencrypted data as a result of carefully crafted Host headers. This is an
-        // extreme edge case but it's potentially exploitable without this check.
-        // DO NOT REMOVE THIS UNLESS YOU'RE SURE YOU KNOW WHAT YOU'RE DOING.
-        if ($vhost->isEncrypted() != $ireq->client->isEncrypted) {
-            return null;
+        $addressPortWildcardHost = "*:0{$serverId}";
```

Comment:
Why that `0` in there?

Reply:
Allow any port.

Reply:
Should probably be `*` then as in the definition of hosts?

Reply:
Nah, makes things easier to keep this as integer. It's anyway just an implementation detail between Vhost and VhostContainer mostly.

Reply:
They should probably both be marked as `@internal`.